server/src/modules/agents/agents.service.ts

```diff
@@ -26,7 +38,51 @@ export class AgentsService {
 
     return {
       ...agent,
-      owner
+      owner: owner.id
     };
   }
+
+  public async updateOne(criteria, partialEntity): Promise<UpdateResult> {
+    return await this.agentRepository.update(criteria, partialEntity);
+  }
+
+  public async trainNlp(id: string, forceUpdate: boolean) {
+    const agent = await this.agentRepository.findOne(id);
+    const options = {
+      languages: this.formatLanguages(agent.languages)
+    };
+    const manager = this.nlpService.initManager(options);
+    const modelPath = this.getModelPath(agent.id);
+
+    this.logger.log(`modelPath -> ${modelPath}`);
+
+    await this.agentRepository.update(id, { status: status.training });
+
+    this.nlpService.train(manager, agent, modelPath, forceUpdate).then(() => {
+      this.manager = manager;
+      this.agentRepository.update(id, { status: status.ready });
+    });
+  }
+
+  public async processLine(id: string, line: string) {
+    const agent = await this.agentRepository.findOne(id);
+
+    if(agent.status === status.dumb) {
+      throw new NoTrainedAgentException();
+    }
+
+    return await this.nlpService.processLine(line, this.manager);
+  }
+
+  private formatLanguages(rawLanguages) {
+    const languages = [];
+
+    rawLanguages.forEach(language => languages.push(language.code));
+
+    return languages;
+  }
+
+  private getModelPath(id) {
+    return path.join(__dirname, '..','..', '..', '..', 'nlp-models',`modelnlp-${id}.nlp`);
```

Comment:
Folder should be a settings parameter / environment variable. This is important because when deploying to a cloud the mounted volume can have a different folder.
In the future, also, we should left open the persistance of models as a service, so can be stored in S3, that means that the storage should be abstracted from this.

Reply:
👍 Totally right, I've updated this part